tests/all.rs

```diff
@@ -1385,3 +1385,33 @@ fn header_size_overflow() {
         err
     );
 }
+
+#[test]
+#[cfg(unix)]
+fn ownership_preserving() {
+    use std::os::unix::prelude::*;
+
+    let td = t!(TempBuilder::new().prefix("tar-rs").tempdir());
+    let rdr = Cursor::new(tar!("ownership.tar"));
+    let mut ar = Archive::new(rdr);
+    ar.set_preserve_ownerships(true);
+
+    if unsafe { libc::getuid() } == 0 {
+        assert!(ar.unpack(td.path()).is_ok());
+        // validate against premade files
+        // iamuid580800001 has this ownership: 580800001:580800000
+        let meta = std::fs::metadata(td.path().join("iamuid580800000")).unwrap();
+        assert_eq!(meta.uid(), 580800000);
+        assert_eq!(meta.gid(), 580800000);
+        let meta = std::fs::metadata(td.path().join("iamuid580800001")).unwrap();
+        assert_eq!(meta.uid(), 580800001);
+        assert_eq!(meta.gid(), 580800000);
+        let meta = std::fs::metadata(td.path().join("iamuid580800002")).unwrap();
+        assert_eq!(meta.uid(), 580800002);
+        assert_eq!(meta.gid(), 580800002);
+    } else {
+        // it's not possible to unpack tar while preserving ownership
+        // without root permissions
+        assert!(ar.unpack(td.path()).is_err());
+    }
```

Comment:
AFAIK this is unlikely to ever get executed, even in CI, so could a test be written that runs in normal user mode as well?

Reply:
> AFAIK this is unlikely to ever get executed, even in CI, so could a test be written that runs in normal user mode as well?


I think it would be very difficult. `chown` itself is a privileged operation, even on Windows. So it would be not possible to test it properly using normal user permissions.

Although, there might be a way to change `gid` on a file that is currently owned by the current effective user (file `uid` == user `euid`). But this kind of test would not cover the `uid` change, so it would be very difficult if not impossible to thoroughly test this functionality under a normal user account.

Reply:
If that is the situation can you arrange for the test to happen with `sudo` on CI?

Also ideally tthis wouldn't check in a `*.tar` file but build it up here so the tar file can be easily changed as well.